draft-ietf-tls-esni.md

```diff
@@ -528,6 +558,57 @@ servers which have opted in, thus minimizing compatibility
 issues. However, there are two scenarios where that does not
 apply, as detailed below.
 
+## Load Balancing Multiple Independent Providers {#multi-cdn}
+
+The DNS is used to independently transfer two linked pieces of
+information to implement ESNI. The server is identified through DNS
+address records and the server's ESNI configuration is specified in a
+ESI typed record that is transported separately. {{publishing-key}}
+requires that these records are coordinated.
+
+The practice of multi-provider load balancing creates a situation
+where it is not possible to reliably coordinate this information. In
+this configuration an alias is used as the server name. The alias is
+implemented as a DNS CNAME record. This provides a level of indirection
+between the domain owner and the hosting provider. When a domain is
+load balanced, the canonical name of the aliased domain potentially
+changes between different providers on a per-request basis. This can
+lead to the use of uncoordinated addressing and key information from
+different providers which in turn will cause connection failures.
+
+In order to prevent this failure the client confirms that the two
+records being used share the same canonical provenance. Specifically,
+the client MUST confirm that the canonical name of the ESNI record
+matches the canonical name of the address record.
+
+Clients that hold mismatched records MUST resolve the conflict by
+obtaining a new ESNI record (or determining its absence) for the
+canonical name of the address record. If the necessary canonical ESNI
+record itself turns out to be an alias this information SHOULD be used
+but is likely to lead to connection failure.
+
+### Example
+
+~~~~
+ AAAA ? www.example.com
+     -> CNAME www.example.com.provider-A.com
+ ESNI ? www.example.com
+     -> CNAME www.example.com.provider-B.com
+ AAAA ? www.example.com.provider-A.com
+     -> AAAA 2001:DB8::AAAA
+ ESNI ? www.example.com.provider-B.com
+     -> "... KEY FOR B ..."
+~~~~
+
+The client determines the canonical names for the www.example.com ESNI
+and AAAA records do not match. It proceeds by obtaining the canonical
+ESNI record that matches the canonical address record it already has.
+
+~~~~
+    ESNI ? www.example.com.provider-A.com
+        -> "... KEY FOR A ..."
+~~~~
```

Comment:
This algorithm seems over-complicated. I.e., there's no reason to query for the ESNI at a different CNAME than A/AAAA. Also, it seems like there is an open case here.

````
Query for ESNI and A/AAAA records.
If A/AAAA is CNAME=X {
   Query for ESNI at X
} else {
   if ESNI is not CNAME {
     success
} else {
     ???
}
```